### PR TITLE
report(details-renderer): support sub-rows within a table

### PR DESCRIPTION
**Pull request by connorjclark**
If a header defines a `mutli: { key: string}` property, the renderer will render each value in `item[key]` within the same column of that row. All other properties (besides label - it doesn't apply) can be set for this "multi header" - anything not set defaults to the outer `heading`.

Note: Originally there was a toplevel `multi: {}` object, which contained all the keys that could be arrays. Turns out that allowing any property of a audit product item to be an array does not complicate the renderer as much as I thought it would, so I removed it.

## Examples

With this PR, nothing yet uses this feature. To see it in use, see the `bundle-analysis` branch (#10064)

image

```
return {
      items,
      headings: [
        {key: 'url', valueType: 'url', multi: {key: 'sources', valueType: 'code'}, label: str_(i18n.UIStrings.columnURL)},
        {key: 'totalBytes', valueType: 'bytes', multi: {key: 'sourceBytes'}, label: str_(i18n.UIStrings.columnSize)},
        {key: 'wastedBytes', valueType: 'bytes', multi: {key: 'sourceWastedBytes'}, label: str_(i18n.UIStrings.columnWastedBytes)},
      ],
    };
```

__

image

```
    /** @type {LH.Audit.Details.OpportunityColumnHeading[]} */
    const headings = [
      {key: 'source', valueType: 'code', multi: {key: 'urls', valueType: 'url'}, label: str_(i18n.UIStrings.columnName)}, // TODO: or 'Source'?
      {key: '_', valueType: 'bytes', multi: {key: 'sourceBytes'}, granularity: 0.05, label: str_(i18n.UIStrings.columnSize)},
      {key: 'wastedBytes', valueType: 'bytes', granularity: 0.05, label: str_(i18n.UIStrings.columnWastedBytes)},
    ];

    return {
      items,
      headings,
      wastedBytesByUrl,
    };
```

### CODE REVIEW

File: lighthouse-core/test/report/html/renderer/details-renderer-test.js [hunk 1]
```diff
@@ -548,7 +548,7 @@ describe('DetailsRenderer', () => {
         // itemType is overriden by code object
         headings: [{key: 'content', itemType: 'url', text: 'Heading'}],
         items: [
-          {content: {type: 'code', value: 'code object'}},
+          {content: {type: 'code', value: 'https://codeobject.com'}},
```
Comment (connorjclark) [inline lighthouse-core/test/report/html/renderer/details-renderer-test.js]: The fallback code in the details render will render a URL w/ an invalid url as a `code` type - so, this test could possible have a false positive (if we ever refactored and broke the fallback type mechanism).

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 2]
```diff
@@ -246,7 +246,7 @@ class DetailsRenderer {
     switch (heading.valueType) {
       case 'bytes': {
         const numValue = Number(value);
-        return this._renderBytes({value: numValue, granularity: 1});
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/details-renderer.js]: this is unrelated to multi, right? just needed for the eventual js audits?

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 3]
```diff
@@ -304,16 +304,45 @@ class DetailsRenderer {
     }
 
     return tableLike.headings.map(heading => {
+      let multi;
+      if (heading.multi) {
+        multi = {
+          key: heading.multi.key,
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/details-renderer.js]: nit: would be nice to reuse this mapping bit rather than manually map since that's what the original point of the function was, but not big deal

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 4]
```diff
@@ -304,16 +304,45 @@ class DetailsRenderer {
...
   }
 
+  /**
+   * @param {LH.Audit.Details.TableItem} row
+   * @param {LH.Audit.Details.OpportunityColumnHeading} heading
+   * @return {Element?}
+   */
+  _renderMultiValue(row, heading) {
+    const values = row[heading.key];
+    if (!Array.isArray(values)) return null;
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/details-renderer.js]: this is always unexpected, right? any recovery options we have?

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 5]
```diff
@@ -304,16 +304,45 @@ class DetailsRenderer {
...
+    const valueElement = this._dom.createElement('div', 'lh-multi-values');
+    for (const childValue of values) {
+      const childValueElement = this._renderTableValue(childValue, heading);
```
Comment (patrickhulce) [inline lighthouse-core/report/html/renderer/details-renderer.js]: does this actually work with arbitrary value types or is report css only setup to accept text/url/code/numbers?

**Comment by connorjclark** [comment]
> bike house : WDYT about subitems instead of multi, I had a difficult time at first understanding why the header definition was shaped the way it was until I got to the bit about valueElement and the ChildElement. subitems, children, something along those lines slightly better reflects the relationship IMO

`subitems` might work

> Does it still complicate protobuf story?

the `details` field is an arbitrary struct, so no change on the encoding side. and we don't do anything specific with it in LR, we just pass it to a json serializer.

File: lighthouse-core/report/html/renderer/details-renderer.js (hunk 2 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js]: yeah I noticed it when I was making the js audits. I could pull this out as a one line change. It seems like a mistake to me.

File: lighthouse-core/report/html/renderer/details-renderer.js (hunk 5 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js]: I'd expect it to work for any `Value` types, although I've only used what you listed so far. For `code`, the default `pre` browser styles had to be tweaked (way too much whitespace). maybe others would need the same.

File: lighthouse-core/report/html/renderer/details-renderer.js (hunk 4 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js]: this is mostly for ts. perhaps it should `console.error` something.

I suppose we could wrap a scalar value in an array of one item - but I would rather be strict about what we accept here. If the heading key was misconfigured, the audit developer will surely notice before publishing (otherwise, the data won't ever show!) Same for if the row value is scalar instead of the expected array. Then there's the case where it, by some ungodly bug, is sometimes scalar, sometimes an array - I bet that won't ever happen, so I think we shouldn't bother with a fallback.

File: lighthouse-core/report/html/renderer/details-renderer.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js]: done, but had to disable some ts.

**Comment by connorjclark** [comment]
how about subheading?

**Comment by patrickhulce** [comment]
> how about subheading?

Oh I'd definitely prefer `multi` over `subheading`. It's kind of a hack that we're using `headings` for this at all because it really has nothing to do with what the heading is, `headings` is really more `columnDefinitions` at this point. Aren't they just basically extra items that are nested in the same row underneath a different, primary item?

**Comment by connorjclark** [comment]
> > how about subheading?
> 
> Oh I'd definitely prefer `multi` over `subheading`. It's kind of a hack that we're using `headings` for this at all because it really has nothing to do with what the heading is, `headings` is really more `columnDefinitions` at this point. Aren't they just basically extra items that are nested in the same row underneath a different, primary item?

ah, that's interesting, to me it seems like a good conceptual fit.

before, each heading defined a single scalar type `(heading.itemType)`. With these changes, it's `(heading.itemType, subheading.itemType*)`. The extra values are displayed vertically so that they _look_ like extra rows, but mentally I consider them to be all one big value in this column's cell. further, all the properties of a `heading` apply (except for `label`).

**Comment by patrickhulce** [comment]
Oh, I mean I agree the thing we currently call `headings` is the perfect fit for this information. I'm just saying it's pretty much a lie that we call it `headings` because it's defining the behavior of the entire column, not just a label for the heading :)

**Comment by paulirish** [comment]
> 🚲 🏠 : WDYT about `subitems` instead of `multi`

big +1. also `multi` being a word fragment instead of a noun makes it extra hard.
but i like `subitems` a lot

and yes at this point it'd make sense to rename `headings` to `columns` but.... we'll have to tackle that separately.

**Comment by connorjclark** [comment]
```
and yes at this point it'd make sense to rename headings to columns but.... we'll have to tackle that separately.
```

that's fine but idk about splitting the terms here at this point. why not keep it consistent (`headings` / `subheadings`), and if we ever rename the former we can rename the latter?

**Comment by patrickhulce** [comment]
> that's fine but idk about splitting the terms here at this point.

We're not splitting the terms, we use `items` for the main items and when we define how the subitems are rendered we use `subitems` :)

The fact you really like `subheadings` makes me feel like I probably misunderstood how this works 😆  

These are multiple values supplied at each individual row, right? AFAICT, we don't have any additional headings as a result of this PR, just additional data being rendered in each row, or is that wrong?

**Comment by connorjclark** [comment]
> we use items for the main items and when we define how the subitems are rendered we use subitems

Both the `.key` and `.multi.key` access the same `item`. I feel `subitems` is misleading.

Paul and I drew some stuff on some paper and came up with a term we both like - `subRows`.

**Comment by paulirish** [comment]
> Paul and I drew some stuff on some paper 


we chatted about this a bit offline. i remembered that a design decision of ours was to make `items` super idiomatic in its design.. basically an ideal object shape for JSON consumers. this is why it's not titled `rows`.

however `headings` is all about the visual presentation in the report.  a more descriptive name would be `columnDefinitions` probably. 

but here i think we can afford to be quite explicit about how this new property will affect the visual representation. so `subRows` works for me.

(yes slightly weird that plurals Rows is not an array, but i also think it works)

**Comment by patrickhulce** [comment]
`subrows` sgtm 👍

File: lighthouse-core/report/html/renderer/details-renderer.js [hunk 6]
```diff
@@ -296,22 +296,54 @@ class DetailsRenderer {
...
+      label: heading.text,
+      displayUnit: heading.displayUnit,
+      granularity: heading.granularity,
+    };
+  }
+
+  /**
+   * @param {LH.Audit.Details.Value[]} values
+   * @param {LH.Audit.Details.OpportunityColumnHeading} heading
+   * @return {Element?}
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/details-renderer.js]: it used to not :)
